### ISSUE TITLE
Mutation update problematic cases with the IN operator in a filtering expression

### ISSUE THREAD

**Issue by healiseu** [yandex/ClickHouse#3597]
Hi, I want to update the value of a single column for thousand-millions-... of rows and mutations is exactly the feature I want to use for this task. I have also read this [passage from documentation][1] but I don't know how this is applied to the following examples.

>The filter_expr must be of type UInt8. This query updates values of specified columns to the values of corresponding expressions in rows for which the filter_expr takes a non-zero value. 

This is the table with some data and the partitions
```sql
CREATE TABLE Test (   
    dim3 UInt16,
    dim2 UInt16,
    dim1 UInt16,

    val  Float32,
    sel  UInt8)
 ENGINE = MergeTree()
 PARTITION BY (dim3, dim2)
 ORDER BY (dim3, dim2, dim1)

┌─dim3─┬─dim2─┬─dim1─┬────val─┬─sel─┐
│ 2222 │  200 │    2 │ 140.15 │   0 │
│ 2222 │  200 │    3 │ 140.28 │   0 │
│ 2222 │  200 │    5 │ 140.33 │   0 │
│ 2222 │  200 │    6 │ 141.62 │   0 │
│ 2222 │  200 │    7 │ 142.61 │   0 │
│ 2222 │  200 │    9 │ 143.64 │   0 │
└──────┴──────┴──────┴────────┴─────┘
┌─dim3─┬─dim2─┬─dim1─┬────val─┬─sel─┐
│ 1222 │  100 │    1 │ 111.45 │   0 │
│ 1222 │  100 │    2 │ 121.15 │   0 │
│ 1222 │  100 │    3 │ 128.46 │   0 │
│ 1222 │  100 │    4 │ 136.15 │   0 │
│ 1222 │  100 │    8 │ 147.77 │   0 │
│ 1222 │  100 │   12 │ 150.02 │   0 │
│ 1222 │  100 │   19 │ 159.01 │   0 │
└──────┴──────┴──────┴────────┴─────┘
┌─dim3─┬─dim2─┬─dim1─┬────val─┬─sel─┐
│ 3333 │  100 │   37 │ 117.77 │   0 │
│ 3333 │  100 │   48 │ 120.02 │   0 │
│ 3333 │  100 │   56 │ 129.01 │   0 │
│ 3333 │  100 │   65 │ 137.01 │   0 │
│ 3333 │  100 │   74 │ 145.01 │   0 │
└──────┴──────┴──────┴────────┴─────┘

OPTIMIZE TABLE 'Test' FINAL

┌─table─┬─database─┬─partition_id─┬─name───────────┬─active─┬─marks─┬─rows─┬─min_blk─┬─max_blk─┬─level─┬─pk_mem─┐
│ Test  │ DemoDB   │ 1222-100     │ 1222-100_1_4_1 │      1 │     1 │    7 │       1 │       4 │     1 │      6 │
│ Test  │ DemoDB   │ 2222-200     │ 2222-200_2_2_1 │      1 │     1 │    6 │       2 │       2 │     1 │      6 │
│ Test  │ DemoDB   │ 3333-100     │ 3333-100_3_3_1 │      1 │     1 │    5 │       3 │       3 │     1 │      6 │
└───────┴──────────┴──────────────┴────────────────┴────────┴───────┴──────┴─────────┴─────────┴───────┴────────┘
```

### Three Working Mutation Examples

```sql
ALTER TABLE Test UPDATE sel=1 WHERE val<130

SELECT * FROM Test  WHERE val < 130

┌─dim3─┬─dim2─┬─dim1─┬────val─┬─sel─┐
│ 1222 │  100 │    1 │ 111.45 │   1 │
│ 1222 │  100 │    2 │ 121.15 │   1 │
│ 1222 │  100 │    3 │ 128.46 │   1 │
└──────┴──────┴──────┴────────┴─────┘
┌─dim3─┬─dim2─┬─dim1─┬────val─┬─sel─┐
│ 3333 │  100 │   37 │ 117.77 │   1 │
│ 3333 │  100 │   48 │ 120.02 │   1 │
│ 3333 │  100 │   56 │ 129.01 │   1 │
└──────┴──────┴──────┴────────┴─────┘

ALTER TABLE Test  UPDATE sel=0 WHERE dim1 IN (37, 48);

SELECT * FROM Test  WHERE dim1 IN (37, 48)

┌─dim3─┬─dim2─┬─dim1─┬────val─┬─sel─┐
│ 3333 │  100 │   37 │ 117.77 │   0 │
│ 3333 │  100 │   48 │ 120.02 │   0 │
└──────┴──────┴──────┴────────┴─────┘

ALTER TABLE Test 
UPDATE sel=1 WHERE [dim3, dim2, dim1] in ([3333, 100, 37], [3333, 100, 48]);

SELECT * FROM Test 
WHERE [dim3, dim2, dim1] IN ([3333, 100, 37], [3333, 100, 48])

┌─dim3─┬─dim2─┬─dim1─┬────val─┬─sel─┐
│ 3333 │  100 │   37 │ 117.77 │   1 │
│ 3333 │  100 │   48 │ 120.02 │   1 │
└──────┴──────┴──────┴────────┴─────┘
```
## Problematic cases for mutations

```sql
ALTER TABLE Test 
UPDATE sel=0 
WHERE [dim3, dim2, dim1] IN (select [dim3, dim2, dim1] from Test where dim1 IN (37, 48));

CREATE VIEW VW AS
SELECT [dim3, dim2, dim1] as vec FROM Test WHERE val<130;

ALTER TABLE Test 
UPDATE sel=0 WHERE [dim3, dim2, dim1] in VW;


SELECT create_time, block_numbers.number AS blk, parts_to_do AS parts, is_done, command
FROM system.mutations WHERE table = 'Test'

┌─────────create_time─┬─blk─┬─parts─┬─is_done─┬─command──────────────────────────────────────────────────────────────────────────────────────────────────┐
│ 2018-11-16 20:41:39 │ [5] │     0 │       1 │ UPDATE sel = 1 WHERE val < 130                                                                           │
│ 2018-11-16 20:43:29 │ [6] │     0 │       1 │ UPDATE sel = 0 WHERE dim1 IN (37, 48)                                                                    │
│ 2018-11-16 20:45:19 │ [7] │     0 │       1 │ UPDATE sel = 1 WHERE [dim3, dim2, dim1] IN ([3333, 100, 37], [3333, 100, 48])                            │
│ 2018-11-16 20:49:50 │ [8] │     3 │       0 │ UPDATE sel = 0 WHERE [dim3, dim2, dim1] IN (SELECT [dim3, dim2, dim1] FROM Test  WHERE dim1 IN (37, 48)) │
│ 2018-11-16 20:51:10 │ [9] │     3 │       0 │ UPDATE sel = 0 WHERE [dim3, dim2, dim1] IN VW                                                            │
└─────────────────────┴─────┴───────┴─────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```
That clearly illustrates that the last two update queries have not been finished. I also noticed that execution of mutations is happening in a queue. If any of these update queries cannot be finished, other workable update queries cannot be finished too. For example if I try again

```sql
ALTER TABLE Test UPDATE sel=1 WHERE val<130


┌─────────create_time─┬─blk──┬─parts─┬─is_done─┬─command──────────────────────────────────────────────────────────────────────────────────────────────────┐
│ 2018-11-16 20:41:39 │ [5]  │     0 │       1 │ UPDATE sel = 1 WHERE val < 130                                                                           │
│ 2018-11-16 20:43:29 │ [6]  │     0 │       1 │ UPDATE sel = 1 WHERE dim1 IN (37, 48)                                                                    │
│ 2018-11-16 20:45:19 │ [7]  │     0 │       1 │ UPDATE sel = 0 WHERE [dim3, dim2, dim1] IN ([3333, 100, 37], [3333, 100, 48])                            │
│ 2018-11-16 20:49:50 │ [8]  │     3 │       0 │ UPDATE sel = 0 WHERE [dim3, dim2, dim1] IN (SELECT [dim3, dim2, dim1] FROM Test  WHERE dim1 IN (37, 48)) │
│ 2018-11-16 20:51:10 │ [9]  │     3 │       0 │ UPDATE sel = 0 WHERE [dim3, dim2, dim1] IN VW                                                            │
│ 2018-11-16 20:58:29 │ [10] │     3 │       0 │ UPDATE sel = 1 WHERE val < 130                                                                           │
└─────────────────────┴──────┴───────┴─────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```
This workable update is added in the queue and it is never finished because of its predecessors.

## Questions
1. Is this behavior of select subquery (View) in the filtering expression of mutations a bug or is it a limitation in current functionality of `ALTER UPDATE field` command ? Recently I discovered that the use of a SET engine with the IN operator is causing the same problem.

2. What is a good alternative method to pass millions of values to the right side of the IN operator for the update of a single column ? 

3. Is there also a possibility to drop the old column and add a new one with the mutated values in place ? This is what I tried without any success
```sql
SELECT if([dim3, dim2, dim1] IN 
    (
        SELECT [dim3, dim2, dim1]
        FROM Test 
        WHERE val < 130
    ), 1, 0) AS sel
FROM Test 

┌─sel─┐
│   0 │
│   0 │
│   0 │
│   0 │
│   0 │
│   0 │
└─────┘
┌─sel─┐
│   1 │
│   1 │
│   1 │
│   0 │
│   0 │
│   0 │
│   0 │
└─────┘
┌─sel─┐
│   1 │
│   1 │
│   1 │
│   0 │
│   0 │
└─────┘

CREATE MATERIALIZED VIEW VWMEM ENGINE=Memory 
POPULATE AS SELECT [dim3, dim2, dim1] as vec 
FROM Test WHERE val<130

ALTER TABLE Test drop column sel;

ALTER TABLE Test
    ADD COLUMN 
    sel UInt8 DEFAULT if([dim3, dim2, dim1] IN DemoDB.VWMEM, 1, 0)

or

ALTER TABLE Test
    ADD COLUMN 
    sel UInt8 DEFAULT if([dim3, dim2, dim1] IN DemoDB.VW, 1, 0)

Ok.

SELECT sel FROM Test WHERE val < 130

┌─sel─┐
│   0 │
│   0 │
│   0 │
└─────┘
┌─sel─┐
│   0 │
│   0 │
│   0 │
└─────┘
```
**Important notice:** This last ALTER command has been accepted from ClickView without any errors or complains but the default values in `sel` column are not those that the if condition calculates. Nevertheless if I use list of arrays in the right side of IN operator instead of (MATERIALIZED)VIEW the DEFAULT expression works fine and I get the result I want. So that is another bug/limited functionality to check out.
```sql
ALTER TABLE Test
    ADD COLUMN 
    sel UInt8 DEFAULT if([dim3, dim2, dim1] IN ([3333, 100, 37], [3333, 100, 48]), 1, 0)

SELECT sel FROM Test  WHERE val < 130

┌─sel─┐
│   1 │
│   1 │
│   0 │
└─────┘
┌─sel─┐
│   0 │
│   0 │
│   0 │
└─────┘
```
That's it fellows I am waiting for your response ;-)

#### My ClickHouse Release
apt-cache policy clickhouse-server
clickhouse-server:
Installed: 18.14.13

[1]: https://clickhouse.yandex/docs/en/query_language/alter/#mutations

**Comment by healiseu** [comment]
Hi Alex (@ztlpn),  thank you for your reply, I confirm that **all my update queries work if I use the full table name**, that includes sub-queries, views, materialized views and sets as well as drop/add column. I have also asked you about the execution of mutations in a queue. I noticed that if any of these update queries cannot be finished, other workable update queries cannot be finished too. Is that normal ;

**Comment by ztlpn** [comment]
Yes, this is by design - mutations are executed strictly in order, so if one of them cannot be finished, all mutations that come after it unfortunately cannot finish too.

**Comment by 5cmpersec** [comment]
Hi @ztlpn @healiseu,

I tried mutation with SET and it is still failed.

```sql
DROP TABLE IF EXISTS default.test;
CREATE TABLE default.test (d Date default today(), i Int8, v Int8) Engine = MergeTree(d, (d,i), 8192);
INSERT INTO default.test (i, v) VALUES (0, 100);
INSERT INTO default.test (i, v) VALUES (1, 11);
DROP TABLE IF EXISTS default.mset;
CREATE TABLE IF NOT EXISTS default.mset (i UInt8) Engine = Set;
INSERT INTO default.mset VALUES (1);
ALTER TABLE default.test UPDATE v = 99 where i in default.mset;
```

Error on clickhouse-server.log
> DB::Exception: Method read is not supported by storage Set

ClickHouse client version 18.14.18.

Am I missing any configurations?

**Comment by ztlpn** [comment]
Hi @5cmpersec, the problem arises because the `i` column is in the index. It is not specific to mutations (try executing `SELECT * FROM test WHERE i IN mset` - you'll get the same error). BTW it has already been reported: #2542

**Comment by ztlpn** [comment]
I'm going to close this issue because the fix for the default database problem has been released in 18.16.0.
